core/src/solution_submission.rs

```diff
@@ -238,15 +259,40 @@ async fn retry_with_gas_price_increase(
         } else {
             Some(BLOCK_TIMEOUT)
         };
-        let result = submit_solution(gas_price, block_timeout).await;
-        if !is_confirm_timeout(&result) || is_last_iteration {
-            return result;
+        let solution_future = submit_solution(gas_price, block_timeout);
+        let deadline_future = wait_for_deadline();
+        futures::pin_mut!(deadline_future);
+        match future::select(solution_future, deadline_future).await {
+            Either::Left((result, _)) => {
+                if !is_confirm_timeout(&result) || is_last_iteration {
+                    return result;
```

Comment:
Before, we were cancelling as soon as the gas price reached the max. If I read the code correctly, we are now not cancelling at all if the gas price became too expensive. I think we should still cancel in that case, but only after the deadline.

Reply:
My reasoning is that in the last iteration the solution_future should never return a confirm timeout because for the last iteration we set the block timeout to `None`. So in the last iteration we will either hit the deadline or not get a timeout on the solution.
This isn't clear in the code though so I should add a comment.